compiler/src/org.graalvm.compiler.core.amd64/src/org/graalvm/compiler/core/amd64/AMD64LIRGenerator.java

```diff
@@ -441,6 +442,20 @@ public Variable emitByteSwap(Value input) {
     }
 
     @Override
+    public Variable emitArrayCompareTo(JavaKind kind1, JavaKind kind2, Value array1, Value array2, Value length1, Value length2) {
+        LIRKind resultKind = LIRKind.value(AMD64Kind.DWORD);
+        RegisterValue raxRes = AMD64.rax.asValue(resultKind);
+        RegisterValue cnt1 = AMD64.rcx.asValue(length1.getValueKind());
+        RegisterValue cnt2 = AMD64.rdx.asValue(length2.getValueKind());
+        emitMove(cnt1, length1);
+        emitMove(cnt2, length2);
+        append(new AMD64ArrayCompareToOp(this, kind1, kind2, raxRes, array1, array2, cnt1, cnt2));
```

Comment:
I think the register selection logic here is broken and also in AMD64ArrayCompareToOp .  C2 uses bound registers and USE_KILL for all the inputs so I think this needs to work differently.  The assembly itself doesn't seem to care about register selection unless there are hidden assumptions in it so I think you could drop the fixed register usage.  If it requires fixed register usage then the assembly should be asserting them.

Reply:
No, it cares; `pcmpestri` does.  And there are asserts.

Reply:
Oh I see them now.  They are about 150 lines into emitCode which seems a bit late.  :)  I would expect them at the beginning and/or in the constructor.  Can you add them in the constructor?

Sorry I edited my comment in between you reply.  We have to use fixed registers like C2 does to get the kill semantics we need anyway.

Reply:
I was thinking about putting them into the constructor but then didn't do because it would be redundant.  And I wanted to keep the code as close to the native one as possible.  If you want I can add an additional assert.